README.md
ReadItLater API
===============

This is a wrapper over [ReadItLater](http://readitlaterlist.com/) API, so you can work with this great service right from your Ruby app.

Installation
============

Sadly it is not still a gem.

Use
===

Once you have installed the lib (future gem) you will have access to the ReadItLater class:

    user = ReadItLater::User.new("username", "password")
    ril = ReadItLater::Api.new("apikey")

    ril.authenticate(user) -> true

    ril.status(user) -> Object with all the status information

    ril.text("http://weblog.rubyonrails.org/2011/11/20/rails-3-1-3-has-been-released") -> will return a text version of the blog post.

TODO
====

* Create a new user
* Get the user's articles list
* Add new articles to the user's list


changelog.md
# Changelog

##  0.0.x - November x, 2011

* Authenticate
* Status
* Get text


spec/spec_helper.rb
$:.unshift File.expand_path('..', __FILE__)
$:.unshift File.expand_path('../../lib', __FILE__)

require 'rspec'
require 'mocha'
require 'fakeweb'
require 'read_it_later'

RSpec.configure do |config|
  config.mock_with :mocha
end


spec/read_it_later/api.rb
require 'spec_helper'

describe ReadItLater::Api do
  before do
  end
end



spec/read_it_later/api_spec.rb
require 'spec_helper'

describe ReadItLater::Api do
  describe "working as an object" do
    it "should handle the apikey" do
      api = ReadItLater::Api.new("apikey")
      api.apikey.should == "apikey"
    end
  end

  describe "work as an object with a user" do
    before :all do
      FakeWeb.allow_net_connect = false
    end

    before do
      @user = ReadItLater::User.new("username", "password")
      @ril = ReadItLater::Api.new("apikey")
    end

    it "authenticate the user with valid data" do
      url = "https://readitlaterlist.com/v2/auth?username=username&password=password&apikey=apikey"
      FakeWeb.register_uri(:get, url, 
                           :body => "200 OK",
                           :status => ["200", "OK"])

      @ril.authenticate(@user).body.should  eql("200 OK")
    end
  end
end


lib/read_it_later.rb
# The ReadItLater module is a wrapper over ReadItLate API
# to work with it as Ruby objects.
# The main idea is to provide a tool as close to the original
# behavior of the API.

# Author::    Marcos Vanetta  (mailto:marcosvanetta@gmail.com)
# Copyright:: Copyright (c) 2011 Marcos Vanetta 
# License::   LGPL

module ReadItLater
  autoload :Api, 'read_it_later/api'
  autoload :Version, 'read_it_later/version'
  autoload :Header,  'read_it_later/header'
  autoload :Response, 'read_it_later/response'
  autoload :Text, 'read_it_later/text'
  autoload :User, 'read_it_later/user'
  autoload :Wrapper, 'read_it_later/wrapper'
end

=begin
class ReadItLater
  include HTTParty
  base_uri 'https://readitlaterlist.com'

  def initialize(username, password, apikey=nil)
    @apikey = apikey.nil? ? get_apikey : apikey
    @username = username
    @passowrd = passowrd
  end

  def self.text(options = {})
    apikey ||= get_apikey
    url = options[:url]
    mode = options[:mode] ? options[:mode] : "less"
    images = options[:images] ? options[:images] : 0
    query = { :url => url, :mode => mode, :images => images, :apikey => apikey}

    bring("/v2/text", :query => query)
  end

  def self.status
    apikey ||= get_apikey
    query = { :apikey => apikey}
    bring('/v2/api', :query => query)
  end

  def self.get_apikey
    "c66gkx0Zpfu44ad8diT9d77P3fAjlp50"
  end

  def create
    query = { :username => @username, :passowrd => @passowrd, :apikey => @apikey}
    bring("/v2/signup", :query =>  query)
  end

  def bring(path, query = {})
    response = if query[:query][:url]
      Text.bring(path, query)
    else
      get(path, queyr)
    end
    wrapper(response)
  end

  def wrapper(response)
    Response.new(response)
  end
end
=end


lib/read_it_later/api.rb
require 'httparty'

# The ReadItLater module is a wrapper over ReadItLate API
# to work with it as Ruby objects.
# The main idea is to provide a tool as close to the original
# behavior of the API.

# Author::    Marcos Vanetta  (mailto:marcosvanetta@gmail.com)
# Copyright:: Copyright (c) 2011 Marcos Vanetta 
# License::   LGPL
module ReadItLater
  class Api

    # Holds the api_key assigned to this ReadItLater instance.
    attr_reader :apikey

    include HTTParty
    base_uri 'https://readitlaterlist.com'

    # Creates a new instance of ReadItLater Wrapper
    #
    # @param [String] The Apikey provide by ReadItLater.com 
    #
    #     @ril = ReadItLater.new("yourapikey")
    #
    def initialize(apikey)
      @apikey = apikey
    end

    # Check if the user credentials are valid.
    # 
    # @param [user] User instance from ReadItLater::User class
    # @return [Boolean] true or false
    #
    #     @user = ReadItLater.new("username", "password")
    #
    #     @ril = ReadItLater.new("yourapikey")
    #
    #     @ril.authenticate(@user) # -> true / false
    #
    def authenticate(user)
      response = self.class.get("/v2/auth", :query => generate_query(user))
      if response.code == 200
        true
      elsif response.code == 401
        false
      else
        raise ArgumentError
      end
    end

    def status(user)
      response = self.class.get("/v2/stats", :query => generate_query(user))
      response.headers
    end

    def text(url)
      Text.bring(:url => url, :apikey => @apikey).body
    end

    def generate_query(user = nil, options = {})
      pre_params = if user
        {:username => user.username,
         :password => user.password,
         :apikey => @apikey}
      else
        { :apikey => @apikey }
      end
      pre_params.merge(options)
    end

    def get(user, options = {})
      response = self.class.get("/v2/get", :query => generate_query(user, options))
      response.to_hash
    end
  end
end


lib/read_it_later/header.rb
module ReadItLater
  class Header
    TYPES = {
      "200" => "Request was successful",
      "400" => "Invalid request, please make sure you follow the documentation for proper syntax",
      "401" => "Username and/or password is incorrect",
      "403" => "Rate limit exceeded, please wait a little bit before resubmitting",
      "503" => "Read It Later's sync server is down for scheduled maintenance"
    }

    AVAILABLE_METHODS = %w(x-limit-user-limit x-limit-user-remaining x-limit-user-reset x-limit-key-limit x-limit-key-remaining x-limit-key-reset)

    def initialize(header)
       @header = header
    end

    def status
      st = @header["status"][0][0..2]
      [st, TYPES[st]]
    end

    def method_missing(sym, *args, &block)
      if sym.to_s =~ /^x\-limit/ && @header.key?(sym)
        @header[sym]
      else
        super
      end
    end

    def respond_to?(sym, include_private=false)
      if sym.to_s =~ /^x\-limit/ && @header.key?(sym)
        true
      else
        super
      end
    end
  end
end


lib/read_it_later/wrapper.rb
require 'httparty'

# The ReadItLater module is a wrapper over ReadItLate API
# to work with it as Ruby objects.
# The main idea is to provide a tool as close to the original
# behavior of the API.

# Author::    Marcos Vanetta  (mailto:marcosvanetta@gmail.com)
# Copyright:: Copyright (c) 2011 Marcos Vanetta 
# License::   LGPL
module ReadItLater
  class Wrapper

    # Holds the api_key assigned to this ReadItLater instance.
    attr_reader :apikey

    include HTTParty
    base_uri 'https://readitlaterlist.com'

    # Creates a new instance of ReadItLater Wrapper
    #
    # @param [String] The Apikey provide by ReadItLater.com 
    #
    #     @ril = ReadItLater.new("yourapikey")
    #
    def initialize(apikey)
      @apikey = apikey
    end

    # Check if the user credentials are valid.
    # 
    # @param [user] User instance from ReadItLater::User class
    # @return [Boolean] true or false
    #
    #     @user = ReadItLater.new("username", "password")
    #
    #     @ril = ReadItLater.new("yourapikey")
    #
    #     @ril.authenticate(@user) # -> true / false
    #
    def authenticate(user)
      response = self.class.get("/v2/auth", :query => generate_query(user))
      if response.code == 200
        true
      elsif response.code == 401
        false
      else
        raise ArgumentError
      end
    end

    def status(user)
      response = self.class.get("/v2/stats", :query => generate_query(user))
      response.headers
    end

    def text(url)
      Text.bring(:url => url, :apikey => @apikey).body
    end

    def generate_query(user = nil, options = {})
      pre_params = if user
        {:username => user.username,
         :password => user.password,
         :apikey => @apikey}
      else
        { :apikey => @apikey }
      end
      pre_params.merge(options)
    end

    def get(user, options = {})
      response = self.class.get("/v2/get", :query => generate_query(user, options))
      response.to_hash
    end
  end
end
